### ISSUE TITLE
EpertGate Nan loss using multi-head classifier

### ISSUE THREAD

**Issue by appledora** [ContinualAI/avalanche#1639]
In my current experiments, I am trying to set up a different dataset ( with different number of classes) as separate tasks/experiences. While trying to train on ExpertGate, I am not sure if I am doing the data processing correctly as I am getting `Nan` loss repeatedly. 

Here's my dataprocessing code: 

```python
for idx, (task_name, num_classes) in enumerate(zip(task_list, classes_per_task)):
    print("task: ", task_name)
    _, _, _, train_dataset, test_dataset = get_dataloaders(task_name, 0.8, 16)
    train_avalanche_dataset = _make_taskaware_classification_dataset(train_dataset)
   
    test_avalanche_dataset = _make_taskaware_classification_dataset(test_dataset)
    train_dataset_list.append(train_avalanche_dataset)
    test_dataset_list.append(test_avalanche_dataset)

ncbm = nc_benchmark(
        train_dataset_list,
        test_dataset_list,
        n_experiences = 100,
        task_labels=True,
        shuffle=False,
        class_ids_from_zero_in_each_exp=True,
        one_dataset_per_exp=True,
        train_transform=None,
        eval_transform=None,
    )
```
However, this is give me an error with missing `task_labels`, during training. 

This is how I am initializing the model and strategy: 

```python
model = ExpertGate(shape=(3, 224, 224),device=device)
model.expert.classifier[6] = MultiHeadClassifier(4096)
cl_strategy =  ExpertGateStrategy(model = model, optimizer = optimizer,
                train_mb_size=256,
                eval_mb_size=128,
                train_epochs=2, 
                ae_lr=1e-3,
                device = device
                )
```

Would love to hear any pointers on this? 
In general, what is the best way to set up dataloaders for my particular setting?

**Comment by appledora** [comment]
btw, I wanted to initially ask on your slack - but the invite link is no longer working.

**Comment by niniack** [comment]
Heya, happy to try and help you out with this. I wrote this strategy for Avalanche a while ago. I'd like to make sure it's usable for you. Hopefully we can squash any bugs, if there is something in the implementation. I might be a bit slow in response, so please bear with me. 

It seems like you're using a custom dataset, were you able to repro this issue with a non-custom dataset? And what optimizer,lr are you using?

**Comment by appledora** [comment]
Hey, @niniack . Thanks for your contributions. 
So, I haven't tried using the non-custom datasets as they don't particularly fit my current objective of using each dataset as a separate task. The custom `get_dataloader` function that I have used here simply returns Pytorch dataloader objects and datasets classes. 
Here's the base `Dataset` class I am using: 

```python
class ImageDataset(Dataset):
    def __init__(self):
        self.data_path = ""
        self.data_name = ""
        self.num_classes = 0
        self.train_transform = None
        self.train_csv_path = ""
        self.image_paths = []
        self.labels = []

    def get_num_classes(self):
        return self.num_classes
    def __getitem__(self, index):
        img_path = self.image_paths[index]
        label = self.labels[index]
        img = Image.open(img_path).convert("RGB")
        if self.train_transform:
            img = self.train_transform(img)
            
        return img, label
    def __len__(self):
        return len(self.image_paths)
    
    @property
    def label_dict(self):
        return {i: self.class_map[i] for i in range(self.num_classes)}
    
    def __repr__(self):
        return f"ImageDataset({self.data_name}) with {self.__len__} instances"
```

```python
def get_dataloader(dataset) : 
    split_size = int(0.8 * len(dataset))
    train_dataset, val_dataset = torch.utils.data.random_split(dataset, [split_size, len(dataset) - split_size])

    train_loader = torch.utils.data.DataLoader(train_dataset, batch_size=batch_size, shuffle=True, drop_last=True,  pin_memory=True)
    val_loader = torch.utils.data.DataLoader(val_dataset, batch_size=batch_size, shuffle=False, drop_last=True,  pin_memory=True)

    return train_loader, val_loader, dataset.get_num_classes()
    ```

**Comment by appledora** [comment]
This is the optimizer I am using: 
```python
model = ExpertGate(shape=(3, 224, 224),device=device)
optimizer = SGD(
        model.expert.parameters(), lr=0.1, momentum=0.9, weight_decay=0.0005
    )
